### ISSUE TITLE
Invalid hanshake error for B.xs on perl 5.21.6+ with DEBUGGING, threads and PERL_POISON enabled

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14574]
Migrated from [rt.perl.org#124038](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124038) (status was 'resolved')

Searchable as RT124038$

**Comment by p5pRT** [comment]
### From perl@profvince.com

#### Created by perl@profvince.com



$ /tmp/perltest/bin/perl5\.21\.6 \-e'require B'
B\.c&#8203;: Invalid handshake key got 0xd200080 needed 0xd280080\, binaries are 
mismatched

As far as I can tell\, this only happens for the B module : I've checked 
a number of other core XS modules \(especially those that define a thread 
local context\) and all of them were loading correctly\.

Once your remove one of \-Dusethreads\, \-DDEBUGGING or \-DPERL\_POISON\, the 
issue disappears\.

The problem appeared somewhere between 5\.21\.5 and 5\.21\.6\, probably at 
the very commit xs\_handshake\(\) was introduced \(I've not run a bisect\)\. 
It is still present in 5\.21\.9\.


Vincent

<details><summary>Perl Info</summary>

```

Flags:
     category=library
     severity=medium
     module=B

Site configuration information for perl 5.21.6:

Configured by vincent at Tue Mar 10 12:16:34 BRT 2015.

Summary of my perl5 (revision 5 version 21 subversion 6) configuration:

   Platform:
     osname=darwin, osvers=13.4.0, archname=darwin-thread-multi-2level
     uname='darwin dhcp-147-65-168-187.impa.br 13.4.0 darwin kernel 
version 13.4.0: wed dec 17 19:05:52 pst 2014; 
root:xnu-2422.115.10~1release_x86_64 x86_64 '
     config_args='-des -Dusedevel -Dusethreads -DDEBUGGING 
-Dman1dir=none -Dman3dir=none -Doptimize=-O0 -DPERL_POISON 
-Dprefix=/tmp/perltest'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=define, usemultiplicity=define
     use64bitint=define, use64bitall=define, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-fno-common -DPERL_DARWIN -DDEBUGGING 
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include',
     optimize='-O0 -DPERL_POISON -g',
     cppflags='-fno-common -DPERL_DARWIN -DDEBUGGING 
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
     ccversion='', gccversion='4.2.1 Compatible Apple LLVM 6.0 
(clang-600.0.57)', gccosandvers=''
     intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, 
doublekind=3
     d_longlong=define, longlongsize=8, d_longdbl=define, 
longdblsize=16, longdblkind=3
     ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' 
-fstack-protector -L/usr/local/lib'
     libpth=/usr/local/lib 
/Library/Developer/CommandLineTools/usr/bin/../lib/clang/6.0/lib 
/Library/Developer/CommandLineTools/usr/lib /usr/lib
     libs=-lgdbm -ldbm -ldl -lm -lutil -lc
     perllibs=-ldl -lm -lutil -lc
     libc=, so=dylib, useshrplib=false, libperl=libperl.a
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup 
-L/usr/local/lib -fstack-protector'



@INC for perl 5.21.6:
     /tmp/perltest/lib/site_perl/5.21.6/darwin-thread-multi-2level
     /tmp/perltest/lib/site_perl/5.21.6
     /tmp/perltest/lib/5.21.6/darwin-thread-multi-2level
     /tmp/perltest/lib/5.21.6
     .


Environment for perl 5.21.6:
     DYLD_LIBRARY_PATH (unset)
     HOME=/Users/vincent
     LANG=C
     LANGUAGE (unset)
     LC_ALL=C
     LD_LIBRARY_PATH=/usr/local/lib:/usr/lib
     LOGDIR (unset)
 
PATH=/Users/vincent/bin:/Users/vincent/perl/builds/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin:/opt/X11/bin:/usr/texbin
     PERLDOC_PAGER=less -R
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

On Tue Mar 10 09&#8203;:51&#8203;:14 2015\, perl@&#8203;profvince\.com wrote&#8203;:
> This is a bug report for perl from perl@&#8203;profvince\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.6\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> $ /tmp/perltest/bin/perl5\.21\.6 \-e'require B'
>  B\.c&#8203;: Invalid handshake key got 0xd200080 needed 0xd280080\, binaries
> are
> mismatched
> 
> As far as I can tell\, this only happens for the B module : I've
> checked
> a number of other core XS modules \(especially those that define a
> thread
> local context\) and all of them were loading correctly\.
> 
> Once your remove one of \-Dusethreads\, \-DDEBUGGING or \-DPERL\_POISON\,
> the
> issue disappears\.
> 
> The problem appeared somewhere between 5\.21\.5 and 5\.21\.6\, probably at
> the very commit xs\_handshake\(\) was introduced \(I've not run a bisect\)\.
> It is still present in 5\.21\.9\.
> 
> 
> Vincent
> 

B\.xs compiland and the interp aren't agreeing on the size the interp struct\, which means they aren't binary compatible/crash risk\. B\.xs is not distributed on CPAN\, and is not upgradable\, so it should have been compiled with the same config\.h and same \-D flags as the core interp\. You can't mix and match the threading and DEBUGGING choices between the interp and XS modules\. B\.xs reports its interp struct size is 0xd20 bytes long\, the interp says its it is supposed to be 0xd28\. 8 bytes longer\. I diagnose this by making a \.i file of the XS module\, a \.i file of av\.c \(or any other core file\)\, run both through gnu indent\. Copy paste the interp struct definition from both prettied \.i into separate files\. Run diff on the 2 separate files containing just the struct definition\. Then look at the diff file and figure out what struct memeber is different\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From bitcard@profvince.com

The B distribution has a darwin hints file that was unconditionally overwriting the OPTIMIZE MakeMaker variable\, discarding PERL\_POISON\. I fixed that in commit 4a33c86d\, which also fixes the same potential issue with PerlIO\-via\.


Vincent

**Comment by p5pRT** [comment]
bitcard@profvince.com - Status changed from 'open' to 'resolved'
